core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -0,0 +1,209 @@
+ */
+public abstract class TransportWriteAction<
+            Request extends ReplicatedWriteRequest<Request>,
+            Response extends ReplicationResponse & ReplicatedWriteResponse
+        > extends TransportReplicationAction<Request, Request, Response> {
+
+    protected TransportWriteAction(Settings settings, String actionName, TransportService transportService,
+            ClusterService clusterService, IndicesService indicesService, ThreadPool threadPool, ShardStateAction shardStateAction,
+            ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver, Supplier<Request> request,
+            String executor) {
+        super(settings, actionName, transportService, clusterService, indicesService, threadPool, shardStateAction, actionFilters,
+                indexNameExpressionResolver, request, request, executor);
+    }
+
+    /**
+     * Called on the primary with a reference to the {@linkplain IndexShard} to modify.
+     */
+    protected abstract WriteResult<Response> onPrimaryShard(Request request, IndexShard indexShard) throws Exception;
+
+    /**
+     * Called once per replica with a reference to the {@linkplain IndexShard} to modify.
+     *
+     * @return the translog location of the {@linkplain IndexShard} after the write was completed or null if no write occurred
+     */
+    protected abstract Translog.Location onReplicaShard(Request request, IndexShard indexShard);
+
+    @Override
+    protected final WritePrimaryResult shardOperationOnPrimary(Request request) throws Exception {
+        final ShardId shardId = request.shardId();
+        IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        IndexShard indexShard = indexService.getShard(shardId.id());
+        WriteResult<Response> result = onPrimaryShard(request, indexShard);
+        return new WritePrimaryResult(request, result.getResponse(), result.getLocation(), indexShard);
+    }
+
+    @Override
+    protected final WriteReplicaResult shardOperationOnReplica(Request request) {
+        final ShardId shardId = request.shardId();
+        IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+        IndexShard indexShard = indexService.getShard(shardId.id());
+        Translog.Location location = onReplicaShard(request, indexShard);
+        return new WriteReplicaResult(indexShard, request, location);
+    }
+
+    /**
+     * Simple result from a write action. Write actions have static method to return these so they can integrate with bulk.
+     */
+    public static class WriteResult<Response extends ReplicationResponse> {
+        private final Response response;
+        private final Translog.Location location;
+
+        public WriteResult(Response response, @Nullable Location location) {
+            this.response = response;
+            this.location = location;
+            // Set the ShardInfo to 0 so we can safely send it to the replicas
```

Comment:
Weird, right? It is there because it is pretty close to where it is in master:
https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java#L215

I think it is worth hunting down if we still need this, but I was getting pretty deep in the block_until_refresh stack when I got here so I just slapped a nasty note on it and moved on. I can look into this in the this PR or we can do another?